commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/UriParser.java

```diff
@@ -455,6 +455,156 @@ public static boolean fixSeparators(final StringBuilder name) {
+            while (reading) {
+                reading = readNonSeparator();
+            }
+        }
+        private void removePreviousElement(int to) throws FileSystemException {
+            if (lastSeparator == 0) {
+                // Previous element is missing
+                throw new FileSystemException("vfs.provider/invalid-relative-path.error");
+            }
+            cursor = lastSeparator - 1;
+            while (readNonSeparator()) {
+                cursor = cursor - 2;
+                if (cursor < 0) {
+                    // Previous element is missing
+                    throw new FileSystemException("vfs.provider/invalid-relative-path.error");
+                }
+            }
+            path.delete(cursor, to);
+            lastSeparator = cursor;
+            this.end = path.length();
+            readSeparator();
+        }
+        private void readToNextSeparator() {
+            boolean reading = true;
+            while (reading) {
+                reading = readNonSeparator();
+            }
+        }
+        private boolean readSeparator() {
+            if (cursor == end) {
+                return false;
+            }
+            if (path.charAt(cursor) == SEPARATOR_CHAR) {
+                cursor++;
+                return true;
+            }
+            if (cursor + 2 >= end) {
+                return false;
+            }
+            String sub = path.substring(cursor, cursor + 3);
+            if (sub.equals("%2f") || sub.equals("%2F")) {
+                cursor = cursor + 3;
+                return true;
+            }
+            return false;
+        }
+        private boolean readDot() {
+            if (cursor == end) {
+                return false;
+            }
+            if (path.charAt(cursor) == '.') {
+                cursor++;
+                return true;
+            }
+            if (cursor + 2 >= end) {
+                return false;
+            }
+            String sub = path.substring(cursor, cursor + 3);
+            if (sub.equals("%2e") || sub.equals("%2E")) {
+                cursor = cursor + 3;
+                return true;
+            }
+            return false;
+        }
+        private boolean consumeSeparator() {
+            int from = cursor;
+            if (readSeparator()) {
+                path.delete(from, cursor);
+                cursor = from;
+                this.end = path.length();
+                return true;
+            }
+            return false;
+        }
+        private boolean readNonSeparator() {
+            if (cursor == end) {
+                return false;
+            }
+            if (path.charAt(cursor) == SEPARATOR_CHAR) {
+                return false;
+            }
+            if (cursor + 2 >= end) {
+                cursor++;
+                return true;
+            }
+            String sub = path.substring(cursor + 1, cursor + 3);
+            if (sub.equals("%2f") || sub.equals("%2F")) {

```

Comment:
Refactor magic string into a constant with a comment containing the fact that 2f is a forward slash.